### PR TITLE
Fix for incorrect usage of detach(), cpu(), to()

### LINKED ISSUE #6214: incorrect usage of detach/cpu/to
## 🐛 Bug

Incorrect use of detach() and cpu() during fixing #4592.

## Please reproduce using the BoringModel

You cannot really.

### To Reproduce

Use following [**BoringModel**](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing) and post here

The fix for #4592 has good intentions but an obvious bug slipped through.

It is easy to understand but hard to test it, so let's rely on common sense:
That is the original code:
```
            if isinstance(output, Result):
                output.detach()
                if self.move_metrics_to_cpu:
                    output.cpu()
```

### Expected behavior

And that supposed to be the correct code, obviously:
```
            if isinstance(output, Result):
                output = output.detach()
                if self.move_metrics_to_cpu:
                    output = output.cpu()
```
### Environment
Any.

### Additional context

I will make a trivial PR for this, I just wanted to create a ticket to actually track the issue.
In theory, the bug is general pattern, but in practice these two lines were the only ones affected in PL.

### PR DESCRIPTION
**Pull request by dvolgyes**
## What does this PR do?

If you use .detach()/to()/cpu(), you need to assign it to a new variable,
it does not change the source variable.

Fixes #6214 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
Great catch !

**Review by tchaton — CHANGES_REQUESTED** [review]
Dear @dvolgyes,

The Result object implements in-place operation. Therefore, you changed won't work.
Therefore, I returned self from to, and cpu functions.

Best,
T.C

**Review by tchaton — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@dvolgyes I have applied this to the remaining codebase, mind checkingng it...

**Comment by dvolgyes** [comment]
I see, thank you for the fix.
In place: I would say it is a very confusing behavior, it looks very pytorch-ish, 
but using a different approach (in place).

**Comment by dvolgyes** [comment]
I looked through, and it looks good to me.
As a remark, the models (nn.Module) are in-place,
so if you call a .cuda() on them, they return themselves,
but there it is not necessary.
I think it is a good practice, because you don't have to keep it in mind
which one works in-place (model, some PL), and which one doesn't (tensors).
So I think the model = model.cpu() is not strictly necessary, but I support the change.

**Comment by dvolgyes** [comment]
Codecov: I don't understand how could it be that low. Something got accidentally disabled? (Jumping from 80 percent to 20%)

**Comment by Borda** [comment]
> Codecov: I don't understand how could it be that low. Something got accidentally disabled? (Jumping from 80 percent to 20%)

it just takes time till all tests are finished stat aggregated, so the faster is base pkg testing and full testing takes about 15-20min

**Comment by dvolgyes** [comment]
I see, thanks for the explanation.
I am not familiar with the setup / internals, and I still don't see how this aggregation works, 
but I can live without the details. :) Thanks again.

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 1]
```diff
@@ -276,7 +276,7 @@ def pre_backward(self, closure_loss: torch.Tensor, should_accumulate: bool, opti
     def model_to_device(self):
         if self.root_device.type == "cuda":
             torch.cuda.set_device(self.root_device)
-        self.model.to(self.root_device)
+        self.model = self.model.to(self.root_device)
```
Comment (rohitgr7) [inline pytorch_lightning/plugins/training_type/ddp.py:279]: for `nn.Module` instances (models), inplace works I think but for tensors it doesn't. this is weird..

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/ddp.py:279]: not really that weird, according to my understanding. moving the tensor to a new device means a new reference to the new memory. 
`nn.Module` is still a python object which always stays in cpu memory, therefore reference never changes, and it contains tensors which are moved the same way as you would move a tensor with `.to()`. 

I'm in favor of reverting the changes to the modules in this PR. The PR should only fix where tensors are moved.

Comment (tchaton) [reply pytorch_lightning/plugins/training_type/ddp.py]: Yes, I agree with @awaelchli. Result objects was doing in-place op, but it is at the same level as a tensor, so it makes sense to return from it.
